lib/invoker.js

```diff
@@ -24,21 +25,63 @@ const invoke = (userFunction) => {
     };
 };
 
-module.exports = (userFunctionUri, port) => {
-    const userFunction = lookUpFunction(userFunctionUri);
+const registerShutdown = (destroyHook, timeout, server, shutdownSignal) => {
+    return shutdownSignal.then(() => {
+        logger('Graceful shutdown started');
+        return new Promise(((resolve, reject) => {
+            server.tryShutdown(() => {
+                executeDestroy(destroyHook, timeout)
+                    .then(() => {
+                        logger('Graceful shutdown completed');
+                        resolve();
+                    }, (err) => {
+                        logger(`Error during $destroy hook execution: ${err.toString()}`);
+                        reject(err);
+                    });
+            });
+        }));
+    });
+};
+
+const executeInit = async (initHook, initTimeout) => {
+    logger('Calling $init hook (if any)');
+    return await guardWithTimeout(initHook, initTimeout)
+        .then(() => logger('$init hook successfully completed (or skipped if none)'));
+};
 
+const executeDestroy = async (destroyHook, timeout) => {
+    logger('Calling $destroy hook (if any)');
+    return await guardWithTimeout(destroyHook, timeout)
+        .then(() => logger('$destroy hook successfully completed (or skipped if none)'));
+};
+
+module.exports = async (userFunctionUri, options, shutdownSignal) => {
     const server = new grpc.Server();
-    server.addService(services.RiffService, {invoke: invoke(userFunction)});
 
-    const shutdown = () => {
-        logger('Graceful shutdown started');
-        server.tryShutdown(() => logger('Graceful shutdown completed'));
-    };
-    process.on('SIGTERM', shutdown);
-    process.on('SIGINT', shutdown);
+    const userFunction = lookUpFunction(userFunctionUri);
+
+    const destroyHook = userFunction.$destroy;
+    const destroyHookTimeout = options.$destroyTimeoutMs || 10000;
+    const shutdownPromise = registerShutdown(destroyHook, destroyHookTimeout, server, shutdownSignal);
+
+    const initHookTimeout = options.$initTimeoutMs || 10000;
+    await executeInit(userFunction.$init, initHookTimeout, server)
```

Comment:
it's odd to mix async/await and promise chains like this in the same function. As is, this `await` doesn't do much because it could instead return the promise chain so long as `.then(() => ({userFunction, shutdownPromise}))` is appended.

I'd recommend flattening the existing promise chain to use async/await with sync control flow. Something like:

```js
try {
    await executeInit(userFunction.$init, initHookTimeout, server);
} catch (err) {
    // TODO trigger shutdown
    return
}
server.addService(services.RiffService, {invoke: invoke(userFunction)});
server.bind(`0.0.0.0:${options.port}`, grpc.ServerCredentials.createInsecure());
server.start();
logger('Ready to process signals');
```

Reply:
As it required less rework, I opted for getting rid of `await` and appending `.then(() => ({userFunction, shutdownPromise}))`. Let me know if that's acceptable.